server/src/main/java/io/crate/types/FloatVectorType.java

```diff
@@ -0,0 +1,203 @@
+            return new FloatVectorIndexer(ref, getFieldType.apply(ref.column()));
+        }
+    };
+
+    private final int dimensions;
+
+    public FloatVectorType(int dimensions) {
+        this.dimensions = dimensions;
+    }
+
+    public FloatVectorType(StreamInput in) throws IOException {
+        this.dimensions = in.readVInt();
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeVInt(dimensions);
+    }
+
+    @Override
+    public int compare(float[] o1, float[] o2) {
+        return Arrays.compare(o1, o2);
+    }
+
+    @Override
+    public int id() {
+        return ID;
+    }
+
+    @Override
+    public Precedence precedence() {
+        return Precedence.CUSTOM;
+    }
+
+    @Override
+    public String getName() {
+        return NAME;
+    }
+
+    @Override
+    public Streamer<float[]> streamer() {
+        return this;
+    }
+
+    @Override
+    public float[] sanitizeValue(Object value) {
+        // Value is stored as list in _raw/xcontent/json
+        if (value instanceof List<?> values) {
+            float[] result = new float[values.size()];
+            for (int i = 0; i < result.length; i++) {
+                result[i] = ((Number) values.get(i)).floatValue();
+            }
+            return result;
+        }
+        return (float[]) value;
+    }
+
+    @Override
+    public float[] implicitCast(Object value) throws IllegalArgumentException, ClassCastException {
+        return sanitizeValue(value);
+    }
+
+    @Override
+    public long valueBytes(float @Nullable [] value) {
+        if (value == null) {
+            return RamUsageEstimator.NUM_BYTES_OBJECT_REF;
+        }
+        return RamUsageEstimator.sizeOf(value);
+    }
+
+    @Override
+    public TypeSignature getTypeSignature() {
+        return new TypeSignature(getName(), List.of(TypeSignature.of(dimensions)));
+    }
+
+    @Override
+    public List<DataType<?>> getTypeParameters() {
+        return List.of(DataTypes.INTEGER);
+    }
+
+    @Override
+    public Integer characterMaximumLength() {
+        return dimensions;
+    }
+
+    @Override
+    public ColumnType<Expression> toColumnType(ColumnPolicy columnPolicy,
+                                               @Nullable Supplier<List<ColumnDefinition<Expression>>> convertChildColumn) {
+        return new ColumnType<>(getName(), List.of(dimensions));
+    }
+
+    @Override
+    public void addMappingOptions(Map<String, Object> mapping) {
+        mapping.put("dimensions", dimensions);
+    }
+
+    @Override
+    public float[] readValueFrom(StreamInput in) throws IOException {
+        if (in.readBoolean()) {
+            return in.readFloatArray();
+        } else {
+            return null;
+        }
+    }
+
+    @Override
+    public void writeValueTo(StreamOutput out, float[] v) throws IOException {
+        if (v == null) {
```

Comment:
minor: can we use :`writeOptionalArray`?

Reply:
`float[]` is not compatible with `T[]`